dev/Microsoft.Process.Environment/EnvironmentManager.cpp

```diff
@@ -63,4 +74,125 @@ namespace winrt::Microsoft::ProjectReunion::implementation
+        THROW_HR_IF_NULL(E_POINTER, environmentVariablesString);
+
+        // parse the block into a map
+        LPTSTR variable;
+        variable = (LPTSTR)environmentVariablesString;
+
+        StringMap environmentVariables;
+        while (*variable)
+        {
+            std::wstring environmentVariable(variable);
+            size_t locationOfNull = environmentVariable.find_last_of(L'=');
+
+            environmentVariables.Insert(environmentVariable.substr(0, locationOfNull), environmentVariable.substr(locationOfNull + 1));
+
+            variable += lstrlen(variable) + 1;
+        }
+
+        THROW_IF_WIN32_BOOL_FALSE(FreeEnvironmentStrings(environmentVariablesString));
+
+        return environmentVariables;
+    }
+
+    StringMap EnvironmentManager::GetUserOrMachineEnvironmentVariables()
+    {
+        StringMap environmentVariables;
+        wil::unique_hkey environmentVariablesHKey = GetRegHKeyForEVUserAndMachineScope();
+
+        // While this way of calculating the max size of the names,
+        // values, and total number of entries includes two calls
+        // to the registry, I believe this is superior to
+        // using a do/while or a while with a prime
+        // because there is no chance of the loop iterating more than
+        // is needed AND the size of the name and value arrays are
+        // only as big as the biggest name or value.
+
+        DWORD sizeOfLongestNameInCharacters;
+        DWORD sizeOfLongestValueInCharacters;
+        DWORD numberOfValues;
+
+        THROW_IF_FAILED(RegQueryInfoKeyW(
+            environmentVariablesHKey.get()
+            , nullptr // lpClass
+            , nullptr // lpcchClass
+            , nullptr // lpReserved
+            , nullptr // lpcSubKeys
+            , nullptr // lpcbMaxSubKeyLen
+            , nullptr // lpcbMaxCLassLen
+            , &numberOfValues //lpcValues
+            , &sizeOfLongestNameInCharacters
+            , &sizeOfLongestValueInCharacters
+            , nullptr // lpcmSecurityDescriptor
+            , nullptr)); // lpftLastWriteTime
+
+        // +1 for null character
+        const DWORD NAME_LENGTH = sizeOfLongestNameInCharacters + 1;
+        const DWORD VALUE_SIZE_IN_BYTES = sizeOfLongestValueInCharacters * sizeof(WCHAR);
+
+        std::unique_ptr<wchar_t[]> environmentVariableName(new wchar_t[NAME_LENGTH]);
+        std::unique_ptr<BYTE[]> environmentVariableValue(new BYTE[VALUE_SIZE_IN_BYTES]);
+
+        for (int valueIndex = 0; valueIndex < numberOfValues; valueIndex++)
+        {
+            DWORD nameLength = NAME_LENGTH;
+            DWORD valueSize = VALUE_SIZE_IN_BYTES;
+            LSTATUS enumerationStatus = RegEnumValueW(environmentVariablesHKey.get()
+                , valueIndex
+                , environmentVariableName.get()
+                , &nameLength
+                , nullptr
+                , nullptr
+                , environmentVariableValue.get()
+                , &valueSize);
```

Comment:
Why
do
you
hate
widescreen
monitors
so
?
<g>

Reply:
![image](https://user-images.githubusercontent.com/2517139/111835820-ae9ccd00-88b2-11eb-89a0-d36fa657fd26.png)

I have a widescreen monitor.  It is easier to see.